Source/Core/VideoCommon/TextureDecoder_Common.cpp

```diff
@@ -434,55 +435,206 @@ static inline u32 DecodePixel_Paletted(u16 pixel, TLUTFormat tlutfmt)
   }
 }
 
+/*  General formula for computing texture offset
+    u16 sBlk = s / blockWidth;
+    u16 tBlk = t / blockHeight;
+    u16 widthBlks = (width / blockWidth) + 1;
+    u32 base = (tBlk * widthBlks + sBlk) * blockWidth * blockHeight;
+    u16 blkS = s & (blockWidth - 1);
+    u16 blkT =  t & (blockHeight - 1);
+    u32 blkOff = blkT * blockWidth + blkS;
+*/
+
+static u8 TexDecoder_LoadTexel4(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 3;
+  u16 tBlk = t >> 3;
+  u16 widthBlks = (imageWidth >> 3) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 5;
+  u16 blkS = s & 7;
+  u16 blkT = t & 7;
+  u32 blkOff = (blkT << 3) + blkS;
+
+  int rs = (blkOff & 1) ? 0 : 4;
+  u32 offset = base + (blkOff >> 1);
+
+  return (*(src + offset) >> rs) & 0xF;
+}
+
+static u8 TexDecoder_LoadTexel8(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 3;
+  u16 tBlk = t >> 2;
+  u16 widthBlks = (imageWidth >> 3) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 5;
+  u16 blkS = s & 7;
+  u16 blkT = t & 3;
+  u32 blkOff = (blkT << 3) + blkS;
+
+  return *(src + base + blkOff);
+}
+
+static u16 TexDecoder_LoadTexel16(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 2;
+  u16 tBlk = t >> 2;
+  u16 widthBlks = (imageWidth >> 2) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 4;
+  u16 blkS = s & 3;
+  u16 blkT = t & 3;
+  u32 blkOff = (blkT << 2) + blkS;
+
+  u32 offset = (base + blkOff) << 1;
+  return *(u16*)(src + offset);
+}
+
+static u32 TexDecoder_LoadTexel32(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 2;
+  u16 tBlk = t >> 2;
+  u16 widthBlks = (imageWidth >> 2) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 5;  // shift by 5 is correct
```

Comment:
To clarify this comment: the shift by 5 is correct because there are two 4x4 blocks, one after the other, the first containing AR and the second containing GB.  It *might* be clearer to instead have the code below use `u32 offset = ((base << 1) + blkOff) << 1`, but I'm not sure that really helps.